crates/wasmtime/src/runtime/types.rs

```diff
@@ -431,6 +431,132 @@ impl RefType {
 
 /// The heap types that can Wasm can have references to.
 ///
+/// # Subtyping Hierarchy
+///
+/// Wasm has three different heap type hierarchies:
+///
+/// 1. Function types
+/// 2. External types
+/// 3. Internal types
+///
+/// Each hierarchy has a top type (the common supertype of which everything else
+/// in its hierarchy is a subtype of) and a bottom type (the common subtype of
+/// which everything else in its hierarchy is supertype of).
+///
+/// ## Function Types Hierarchy
+///
+/// The top of the function types hierarchy is `func`; the bottom is
+/// `nofunc`. In between are all the concrete function types.
+///
+/// ```text
+///                          func
+///                       /  /  \  \
+///      ,----------------  /    \  -------------------------.
+///     /                  /      \                           \
+///    |              ,----        -----------.                |
+///    |              |                       |                |
+///    |              |                       |                |
+/// (func)    (func (param i32))    (func (param i32 i32))    ...
+///    |              |                       |                |
+///    |              |                       |                |
+///    |              `---.        ,----------'                |
+///     \                  \      /                           /
+///      `---------------.  \    /  ,------------------------'
+///                       \  \  /  /
+///                         nofunc
+/// ```
+///
+/// Additionally, some concrete function types are sub- or supertypes of other
+/// concrete function types. For simplicity, this isn't depicted in the diagram
```

Comment:
Wait sort of unrelated question on this, in the type section you can explicitly list the supertype and there's also `is_final` which say "no one else can subtype this". With the logic you're explaining below though that's not quite right is it? An `is_final` function type can still be subtyped so long as you align the parameters correctly?

(totally unrelated to this PR I'm just trying to reconcile two different ways I've understood how subtyping is done in wasm)

Reply:
Yeah honestly I don't fully understand how all the dots connect in the way that the spec is written/defined, I just know what the intended behavior is.

FWIW, we don't supertypes and finality yet, but this code (and I suppose these docs) will need to be updated at that point.

Reply:
> With the logic you're explaining below though that's not quite right is it? An `is_final` function type can still be subtyped so long as you align the parameters correctly?

And to be clear, yeah you can't subtype something (including a func type) that is declared final.